README.md
[![Build Status](https://travis-ci.org/rgsmolin/play-sound-mplayer.svg?branch=master)](https://travis-ci.org/rgsmolin/play-sound-mplayer)
[![NPM](https://img.shields.io/npm/v/play-sound-mplayer.svg)](https://www.npmjs.com/package/play-sound-mplayer)
[![Coverage Status](https://coveralls.io/repos/github/rgsmolin/play-sound-mplayer/badge.svg?branch=develop)](https://coveralls.io/github/rgsmolin/play-sound-mplayer?branch=develop)

# play-sound-mplayer

Play sounds by shelling out with MPlayer audio player. Make it easy to play audio in Node.js.

## Installation

``` bash
npm install play-sound-mplayer
```

Since we use  mplayer over the command-line to facilitate audio playback, you will also need to install a mplayer instance.

## Usage

### Getting Started

To get started, require the module in your program and create a new AudioPlayer instance.

```javascript
const mplayer = require('play-sound-mplayer');
const player = new mplayer.AudioPlayer();
```

### Play

To play an audio source, use the `play` function and pass in the file path  or url of the audio source that is to be played:

```javascript
player.on('start', () => {
  // Do something here when the audio starts playing
});

player.on('end', () => {
  // Do something here when the audio finishes playing
});

player.on('error', (error: Error) => {
  // Do something here when an error thrown
});

player.play('./test/sound.mp3');
```

### Stop

Stop the playback (without the ability to resume later) by using the `stop` function:

```javascript
player.on('stop', () => {
  // Do something here when the audio is stopped
});

player.stop();
```

### Pause

Pause the playback (with the ability to resume later) by using the `pause` function:

```javascript
player.on('pause', () => {
  // Do something here when the audio is paused
});

player.pause();
```

### Resume

Resume the playback from a paused state by using the `resume` function:

```javascript
player.on('resume', () => {
  // Do something here when the audio is resumed
});

player.resume();
```

### Mute

Mute the playback (with the ability to unmute later) by using the `mute` function:

```javascript
player.on('muted', () => {
  // Do something here when the audio is muted.
});

player.mute();
```

### Unmute

Unmute the playback from a muted state by using the `unMute` function:

```javascript
player.on('unmuted', () => {
  // Do something here when the audio is unmuted
});

player.unMute();
```

### Set Playback Volume

Set the volume of the playback by using the `setVolume(0..100)` function:

```javascript
player.on('volumeupdate', (volume: number) => {
  // Do something here when the audio volume updated with volume: number 0..100
});

player.setVolume(30);
```


CHANGELOG.md
# [0.1.2](https://github.com/rgsmolin/play-sound-mplayer/compare/master...master) (2019-04-03)

## Fixes

* **errorhandler improved** add handling for invalid streams names

# [0.1.1](https://github.com/rgsmolin/play-sound-mplayer/compare/master...master) (2019-04-03)

## Fixes

* **errorhandler improved** add handling for invalid source files and http 400/ http 403 errors

# [0.1.0](https://github.com/rgsmolin/play-sound-mplayer/compare/master...master) (2019-03-31)

## Features

* **changelog added**
* **debug parameter:** player.play(src, {}, true );
* **custom errorhandler** uncaugth errors from stderr emits a custom error message

test/AudioPlayer.spec.ts
import 'mocha';
import { expect } from 'chai';
import { AudioPlayer} from '../lib/AudioPlayer';

describe('AudioPlayer Test Suite', () => {

    let subject: any;

    describe('Basic Interface Tests', () => {
        beforeEach(function () {
            // runs before each test in this block
            subject = new AudioPlayer();
        });

        it('isPlaying should be false before play an audio source', () => {
            expect(subject.isPlaying).to.equal(false);
        });
        it('isPaused should be false before play an audio source', () => {
            expect(subject.isPaused).to.equal(false);
        });
        it('isMuted should be false before play an audio source', () => {
            expect(subject.isMuted).to.equal(false);
        });

        it('play emits an error when it is not passed an audio source', () => {
            try {
                subject.play('');
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('No audio source specified');
            }
        });
        it('stop emits an error when no audio source is playing', () => {
            try {
                subject.stop();
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('No audio source to stop');
            }
        });
        it('pause emits an error when no audio source is playing', () => {
            try {
                subject.pause();
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('No audio source to pause');
            }
        });
        it('resume emits an error when no audio source is playing', () => {
            try {
                subject.resume();
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('No audio source to resume');
            }
        });
        it('mute emits an error when no audio source is playing', () => {
            try {
                subject.mute();
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('No audio source to mute');
            }
        });
        it('unMute emits an error when no audio source is playing', () => {
            try {
                subject.unMute();
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('No audio source to unmute');
            }
        });
        it('setVolume emits an error hhen it is not passed an audio source', () => {
            try {
                subject.setVolume(NaN);
            } catch (error) {
                // tslint:disable-next-line:no-unused-expression
                expect(error).to.exist;
                expect(error.message).to.equal('Invalid volume argument, should between 0 and 100');
            }
        });
        it('setVolume < 0 results a currentVolume of 0', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(-1);
            expect(subject.currentVolume).to.equal(0);
        });
        it('setVolume > 100 results a currentVolume of 100', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(1000);
            expect(subject.currentVolume).to.equal(100);
        });
        it('setVolume > 100 results a currentVolume of 100', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(1000);
            expect(subject.currentVolume).to.equal(100);
        });
        it('setVolume 30 results a currentVolume of 30', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(30);
            expect(subject.currentVolume).to.equal(30);
        });
    });

   xdescribe('Player Events Tests', () => {
        before(function () {
            // runs before each test in this block
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.mute();
        });
        after(function () {
            // runs before each test in this block
            subject.stop();
        });
        it('start playing should emit a start event', (done) => {
            subject.on('start', () => {
                done();
            });
        });
        it('ending playing should emit a end event', (done) => {
            subject.on('end', () => {
                done();
            });
        });
        xit('stop playing should emit a stop event', (done) => {
            subject.on('stop', () => {
                done();
            });
            subject.stop();
        });

    });
    xdescribe('Player Single Events Tests', () => {
        afterEach(function () {
            // runs before each test in this block
            subject.stop();
        });
        it('mute playing should emit a mute event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('mute', () => {
                done();
            });
            subject.mute();
        });
        it('unmute playing should emit a unmute event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('unmute', () => {
                done();
            });
            subject.mute();
            subject.unMute();
        });

        it('pause playing should emit a pause event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('pause', () => {
                done();
            });
            subject.pause();
        });
        it('resume playing should emit a resume event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.pause();
            subject.on('resume', () => {
                done();
            });
            subject.resume();
        });
        it('setVolume should emit a volumeupdate event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('volumeupdate', () => {
                done();
            });
            subject.setVolume(40);
        });

    });
});


test/audio-player.spec.ts
import 'mocha';
import { expect } from 'chai';
import { AudioPlayer} from '../lib/audio-player';

describe('AudioPlayer Test Suite', () => {

    let subject: any;

    describe('Basic Interface Tests', () => {
        beforeEach(function () {
            // runs before each test in this block
            subject = new AudioPlayer();
        });

        it('isPlaying should be false before play an audio source', () => {
            expect(subject.isPlaying).to.equal(false);
        });

        it('isPaused should be false before play an audio source', () => {
            expect(subject.isPaused).to.equal(false);
        });

        it('isMuted should be false before play an audio source', () => {
            expect(subject.isMuted).to.equal(false);
        });

        it('play emits an error when it is not passed an audio source', () => {
            expect(() => { subject.play(''); }).to.throw(/No audio source specified/);
        });

        it('stop emits an error when no audio source is playing', () => {
            expect(() => { subject.stop(); }).to.throw(/No audio source to stop/);
        });

        it('pause emits an error when no audio source is playing', () => {
            expect(() => { subject.pause(); }).to.throw(/No audio source to pause/);
        });

        it('resume emits an error when no audio source is playing', () => {
            expect(() => { subject.resume(); }).to.throw(/No audio source to resume/);
        });

        it('mute emits an error when no audio source is playing', () => {
            expect(() => { subject.mute(); }).to.throw(/No audio source to mute/);
        });

        it('unMute emits an error when no audio source is playing', () => {
            expect(() => { subject.unMute(); }).to.throw(/No audio source to unmute/);
        });

        it('setVolume emits an error hhen it is not passed an audio source', () => {
            expect(() => { subject.setVolume(NaN); }).to.throw(/Invalid volume argument, should between 0 and 100/);
        });

        it('setVolume < 0 results a currentVolume of 0', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(-1);
            expect(subject.currentVolume).to.equal(0);
        });
        it('setVolume > 100 results a currentVolume of 100', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(1000);
            expect(subject.currentVolume).to.equal(100);
        });
        it('setVolume > 100 results a currentVolume of 100', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(1000);
            expect(subject.currentVolume).to.equal(100);
        });
        it('setVolume 30 results a currentVolume of 30', () => {
            expect(subject.currentVolume).to.equal(100);
            subject.setVolume(30);
            expect(subject.currentVolume).to.equal(30);
        });
    });

   xdescribe('Player Start Tests', () => {
        beforeEach(function () {
            // runs before each test in this block
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.mute();
        });

        afterEach(function () {
            // runs before each test in this block
            subject.stop();
        });

        it('start playing should emit a start event', (done) => {
            subject.on('start', () => {
                done();
            });
        });

        it('ending playing should emit a end event', (done) => {
            subject.on('end', () => {
                done();
            });
        });
    });
    xdescribe('Player Stop Tests', () => {
        it('stop playing should emit a stop event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.mute();
            subject.on('stop', () => {
                done();
            });
            subject.stop();
        });

    });

    xdescribe('Player Single Events Tests', () => {
        afterEach(function () {
            // runs before each test in this block
            subject.stop();
        });
        it('mute playing should emit a mute event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('mute', () => {
                done();
            });
            subject.mute();
        });
        it('unmute playing should emit a unmute event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('unmute', () => {
                done();
            });
            subject.mute();
            subject.unMute();
        });

        it('pause playing should emit a pause event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('pause', () => {
                done();
            });
            subject.pause();
        });
        it('resume playing should emit a resume event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.pause();
            subject.on('resume', () => {
                done();
            });
            subject.resume();
        });
        it('setVolume should emit a volumeupdate event', (done) => {
            subject = new AudioPlayer();
            subject.play('./test/sound.mp3', {});
            subject.on('volumeupdate', () => {
                done();
            });
            subject.setVolume(40);
        });

    });
    xdescribe('Player Errorhandler Tests', () => {
        it('start playing with invalid source file should emit an error event', (done) => {
            subject = new AudioPlayer();
            subject.play('x./test/sound.mp3', { }, true);
            subject.on('error', (error: Error) => {
                done();
            });
        });

        it('start playing with invalid source url should emit an stream not found error event', (done) => {
            subject = new AudioPlayer();
            subject.play('xhttps://xxaudiblecdn-vh.akamaihd.net', {}, true);
            subject.on('error', (error: Error) => {
                done();
            });
        });


        it('start playing with invalid source url should emit an http 400 bad request error event', (done) => {
            subject = new AudioPlayer();
            subject.play('https://xaudiblecdn-vh.akamaihd.net', {}, true);
            subject.on('error', (error: Error) => {
                done();
            });
        });

        it('start playing with invalid source url should emit an http 403 forbidden error event', (done) => {
            subject = new AudioPlayer();
            subject.play('https://xaudiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/argo/002149/V$', {}, true);
            subject.on('error', (error: Error) => {
                done();
            });
        });
    });
});


lib/AudioPlayer.ts
import { spawn, ChildProcess } from 'child_process';
import { EventEmitter } from 'events';

interface IAudioPlayer {
    readonly isPlaying: boolean;
    readonly isPaused: boolean;
    readonly isMuted: boolean;
    readonly currentVolume: number;

    play(path: string, options?: any): void;
    pause(): void;
    resume(): void;
    stop(): void;
    mute(): void;
    unMute(): void;
    setVolume(value: number): void;
}

class AudioPlayer extends EventEmitter implements IAudioPlayer {
    private static readonly KEYWORD_PROGRESS = 'A:';
    private static readonly KEYWORD_STARTING = 'Starting playback...';
    private static readonly KEYWORD_EXITING = 'Exiting...';

    private _player = 'mplayer';
    private _audioProcess: ChildProcess | null = null;

    private _isPlaying = false;
    private _isPaused = false;
    private _isMuted = false;
    private _currentVolume = 100;

    constructor() {
        super();
    }

    public get isPlaying(): boolean {
        return this._isPlaying;
    }
    public get isPaused(): boolean {
        return this._isPaused;
    }
    public get isMuted(): boolean {
        return this._isMuted;
    }
    public get currentVolume(): number {
        return this._currentVolume;
    }

    public play(path: string, options?: any): void {
        options = typeof options === 'object' ? options : {};
        options.stdio = ['pipe', 'pipe', 'pipe'];
        options.shell = false;  // required for stopping
        if (!path) {
            this.emit('error', new Error('No audio source specified'));
        } else {
            this._isPlaying = true;
            this.handlePlay(path, options);
        }
    }
    public stop(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to stop'));
        } else {
            console.log('player stopping');
            this.emit('stop');
            this._audioProcess.kill();
            this._audioProcess = null;
        }
     }
    public pause(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to pause'));
        } else {
            if (!this._isPaused) {
                console.log('player pausing');
                this.emit('pause');
                this._isPaused = true;
                this._audioProcess.stdin.write('pause\n');
            }
        }
    }
    public resume(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to resume'));
        } else {
            if (this._isPaused) {
                console.log('player resuming');
                this._isPaused = false;
                this._audioProcess.stdin.write('pause\n');
                this.emit('resume');
            }
        }
    }

    public mute(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to mute'));
        } else {
            if (!this._isMuted) {
                console.log('player muting');
                this.emit('mute');
                this._isMuted = true;
                this._audioProcess.stdin.write('mute\n');

            }
        }
    }
    public unMute(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to unmute'));
        } else {
            if (this._isMuted) {
                console.log('player unmuting');
                this.emit('unmute');
                this._isMuted = true;
                this._audioProcess.stdin.write('mute\n');

            }
        }
    }
    public setVolume(volRel: number): void {
        if (!volRel) {
            this.emit('error', new Error('Invalid volume argument, should between 0 and 100'));
        }
        if (volRel < 0) {
            volRel = 0;
        }
        if (volRel > 100) {
            volRel = 100;
        }
        this._currentVolume = volRel;
        if (this._audioProcess) {
            this.emit('volumeupdate');
            this._audioProcess.stdin.write('set_property volume ' + this._currentVolume + '\n');
        }
    }
    private reset(): void {
        this._currentVolume = 100;
        this._isPlaying = false;
        this._isMuted = false;
        this._isPaused = false;
    }

    private handlePlay(path: string, options: any): void {
        const args: string[] = ['-slave', path];
        this._audioProcess = spawn(this._player, args, options);

        const rexStart = new RegExp(AudioPlayer.KEYWORD_STARTING);
        const rexEnd = new RegExp(AudioPlayer.KEYWORD_EXITING);

        this._audioProcess.stdout.on('data', (chunk: any) => {
            const output: string = chunk.toString();
            if (output.substr(0, 2) === AudioPlayer.KEYWORD_PROGRESS) {
                // console.log('player progress', output);
                this.emit('progress', output);
            } else {
                if (output.match(rexStart)) {
                    // console.log('player starting');
                    this.emit('start');
                } else if (output.match(rexEnd)) {
                    // console.log('player ending');
                    this.emit('end');
                }
            }
        });

        this._audioProcess.on('exit', (code: number | null, signal: string | null) => {
            // console.log('child process exited with ' + 'code ${code} and signal ${signal}');
            this.reset();
            this.emit('exit');
        });

        this._audioProcess.on('close', (code: number, signal: string) => {
            // console.log('child process closed with ' + 'code ${code} and signal ${signal}');
            this.reset();
            this.emit('close');
        });

        this._audioProcess.on('message', (msg: any)  => {
            // console.log('child process message event ', msg);
            this.emit('message', msg);
        });

        this._audioProcess.on('error', (err: Error) => {
            // console.log('child process error', err);
            this.emit('error', err);
        });

    }
}

// Export the module
module.exports = AudioPlayer;


lib/audio-player.ts
import { spawn, ChildProcess } from 'child_process';
import { EventEmitter } from 'events';

interface IAudioPlayer {
    readonly isPlaying: boolean;
    readonly isPaused: boolean;
    readonly isMuted: boolean;
    readonly currentVolume: number;

    play(path: string, options?: any): void;
    pause(): void;
    resume(): void;
    stop(): void;
    mute(): void;
    unMute(): void;
    setVolume(value: number): void;
}

export class AudioPlayer extends EventEmitter implements IAudioPlayer {
    private static readonly KEYWORD_PROGRESS = 'A:';
    private static readonly KEYWORD_STARTING = 'Starting playback...';
    private static readonly KEYWORD_EXITING = 'Exiting...';

    private _player = 'mplayer';
    private _audioProcess: ChildProcess | null = null;

    private _isPlaying = false;
    private _isPaused = false;
    private _isMuted = false;
    private _currentVolume = 100;

    private _isDebug = false;

    constructor() {
        super();
    }

    public get isPlaying(): boolean {
        return this._isPlaying;
    }
    public get isPaused(): boolean {
        return this._isPaused;
    }
    public get isMuted(): boolean {
        return this._isMuted;
    }
    public get currentVolume(): number {
        return this._currentVolume;
    }

    public play(path: string, options?: any, debug: boolean = false): void {
        options = typeof options === 'object' ? options : {};
        this._isDebug = debug;
        options.stdio = ['pipe', 'pipe', 'pipe'];
        options.shell = false;  // required for stopping
        if (!path) {
            this.emit('error', new Error('No audio source specified'));
        } else {
            this._isPlaying = true;
            this.handlePlay(path, options);
        }
    }
    public stop(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to stop'));
        } else {
            // console.log('player stopping');
            this.emit('stop');
            this._audioProcess.kill();
            this._audioProcess = null;
        }
     }
    public pause(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to pause'));
        } else {
            if (!this._isPaused) {
                // console.log('player pausing');
                this.emit('pause');
                this._isPaused = true;
                this._audioProcess.stdin.write('pause\n');
            }
        }
    }
    public resume(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to resume'));
        } else {
            if (this._isPaused) {
                // console.log('player resuming');
                this._isPaused = false;
                this._audioProcess.stdin.write('pause\n');
                this.emit('resume');
            }
        }
    }

    public mute(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to mute'));
        } else {
            if (!this._isMuted) {
                // console.log('player muting');
                this.emit('mute');
                this._isMuted = true;
                this._audioProcess.stdin.write('mute\n');

            }
        }
    }
    public unMute(): void {
        if (!this._audioProcess) {
            this.emit('error', new Error('No audio source to unmute'));
        } else {
            if (this._isMuted) {
                // console.log('player unmuting');
                this.emit('unmute');
                this._isMuted = true;
                this._audioProcess.stdin.write('mute\n');
            }
        }
    }
    public setVolume(volRel: number): void {
        if (!volRel) {
            this.emit('error', new Error('Invalid volume argument, should between 0 and 100'));
        }
        if (volRel < 0) {
            volRel = 0;
        }
        if (volRel > 100) {
            volRel = 100;
        }
        this._currentVolume = volRel;
        if (this._audioProcess) {
            this.emit('volumeupdate');
            this._audioProcess.stdin.write('set_property volume ' + this._currentVolume + '\n');
        }
    }

    private reset(): void {
        this._currentVolume = 100;
        this._isPlaying = false;
        this._isMuted = false;
        this._isPaused = false;
    }

    private handlePlay(path: string, options: any): void {
        let buffer = new Buffer('');
        const args: string[] = ['-slave', path];
        this._audioProcess = spawn(this._player, args, options);

        const rexStart = new RegExp(AudioPlayer.KEYWORD_STARTING);
        const rexEnd = new RegExp(AudioPlayer.KEYWORD_EXITING);


        this._audioProcess.on('close', (code: number, signal: string) => {
            this.reset();
            // check for errors on stderr
            const customError = this.customError('MPlayerError', buffer);
            if (customError) {
                this.logger('child process error');
                this.emit('error', customError);
            } else {
                this.logger('child process closed with code:' + code + ' and signal:' + signal);
                this.emit('close');
            }
        });

        this._audioProcess.on('error', (err: Error) => {
            this.logger('child process error' + err);
            this.reset();
            this.emit('error', err);
        });

        this._audioProcess.on('exit', (code: number | null, signal: string | null) => {
            this.reset();
            if (buffer.length === 0) {
                this.logger('child process exited with code:' + code + ' and signal:' + signal);
                this.emit('exit');
            }
        });

        this._audioProcess.on('message', (msg: any)  => {
            this.logger('child process message event: ' + msg);
            this.emit('message', msg);
        });

        this._audioProcess.stdout.on('data', (chunk: any) => {
            const output: string = chunk.toString();
            if (output.substr(0, 2) === AudioPlayer.KEYWORD_PROGRESS) {
                // this.logger('player progress:' + output);
                this.emit('progress', output);
            } else {
                if (output.match(rexStart)) {
                    this.logger('player starting');
                    this.emit('start');
                } else if (output.match(rexEnd) && buffer.length === 0) {
                    this.logger('player ending');
                    this.emit('end');
                }
            }
        });

        this._audioProcess.stderr.on('data', (chunk: any) => {
            buffer += chunk;
            // console.log('from stderr', buffer.toString());
        });
    }

    private logger(message: string) {
        if (this._isDebug) {
            console.log(message);
        }
    }

    private customError(name: string, buffer: Buffer): Error | null {
        let retValue = null;
        let message = '';
        if (buffer.length > 0) {
            const bufferStr = buffer.toString();
            this.logger('Buffer:' + bufferStr);
            if (bufferStr.match('File not found')) {
                message = 'File not found error';
            } else if (bufferStr.match('No stream found to handle')) {
                message = 'No stream found to handle';
            }  else if (bufferStr.match('HTTP error 400 Bad Request')) {
                message = ' Http error 400 Bad Request';
            } else if (bufferStr.match('HTTP error 403 Forbidden')) {
                message = ' Http error 403 Forbidden';
            }
            const err = new Error(message);
            err.stack = '';
            retValue = message.length > 0 ? err : null;
        }
        return retValue;
    }
}


